sklearn/linear_model/sag_fast.pyx

```diff
@@ -32,62 +27,270 @@ cdef inline double fmax(double x, double y) nogil:
     return y
 
 
+cdef double _logsumexp(double* arr, int n_classes) nogil:
+    """Computes the sum of arr assuming arr is in the log domain.
+
+    Returns log(sum(exp(arr))) while minimizing the possibility of
+    over/underflow.
+    """
+    # Use the max to normalize, as with the log this is what accumulates
+    # the less errors
+    cdef double vmax = arr[0]
+    cdef double out = 0.0
+    cdef int i
+
+    for i in range(1, n_classes):
+        if vmax < arr[i]:
+            vmax = arr[i]
+
+    for i in range(n_classes):
+        out += exp(arr[i] - vmax)
+
+    return log(out) + vmax
+
+
+cdef class MultinomialLogLoss:
+    cdef double _loss(self, double* prediction, double y, int n_classes,
+                      double sample_weight) nogil:
+        """Multinomial Logistic regression loss.
+
+        The multinomial logistic loss for one sample is:
+        loss = - sw \sum_c \delta_{y,c} (prediction[c] - logsumexp(prediction))
+             = sw (logsumexp(prediction) - prediction[y])
```

Comment:
I find it more explicit to reuse the formula of the reference and to simplify only in the second line.
wdyt?